### ISSUE TITLE
Google OAuth token expiration

### ISSUE THREAD

**Issue by vitaliy-sk** [joeyates/imap-backup#94]
Seems Gmail OAuth authenticator not refreshing the OAuth token.
After expiration of the access token it throwing an exception

```
/ruby/lib/ruby/gems/3.0.0/gems/signet-0.14.1/lib/signet/oauth_2/client.rb:1039:in `fetch_access_token': Authorization failed.  Server message: (Signet::AuthorizationError)
{
  "error": "invalid_grant",
  "error_description": "Token has been expired or revoked."
}
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/signet-0.14.1/lib/signet/oauth_2/client.rb:1054:in `fetch_access_token!'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/googleauth-0.15.1/lib/googleauth/signet.rb:81:in `block in fetch_access_token!'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/googleauth-0.15.1/lib/googleauth/signet.rb:108:in `retry_with_error'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/googleauth-0.15.1/lib/googleauth/signet.rb:80:in `fetch_access_token!'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/signet-0.14.1/lib/signet/oauth_2/client.rb:1069:in `refresh!'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/imap-backup-3.2.0/lib/gmail/authenticator.rb:93:in `block in credentials'
	from <internal:kernel>:90:in `tap'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/imap-backup-3.2.0/lib/gmail/authenticator.rb:92:in `credentials'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/imap-backup-3.2.0/lib/imap/backup/account/connection.rb:96:in `block in imap'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/imap-backup-3.2.0/lib/retry_on_error.rb:4:in `retry_on_error'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/imap-backup-3.2.0/lib/imap/backup/account/connection.rb:88:in `imap'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/imap-backup-3.2.0/lib/imap/backup/account/connection.rb:61:in `run_backup'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/imap-backup-3.2.0/lib/imap/backup/configuration/list.rb:20:in `block in each_connection'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/imap-backup-3.2.0/lib/imap/backup/configuration/list.rb:18:in `each'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/imap-backup-3.2.0/lib/imap/backup/configuration/list.rb:18:in `each_connection'
	from /volume1/@appstore/ruby/lib/ruby/gems/3.0.0/gems/imap-backup-3.2.0/bin/imap-backup:76:in `<main>'
```

**Comment by joeyates** [comment]
Hi @vitaliy-sk Thanks for the report.

The error looks like the failure is happening during an attempt to refresh the access token.

I've done a search and it looks like there can be many reasons for failed renewal (too may tokens, the token has actually been revoked manually, password change).

Do any of these cases seem to match your situation?

By the way, I've released version 3.3.0, which makes repeating access token setup a bit easier.

**Comment by vitaliy-sk** [comment]
> I've done a search and it looks like there can be many reasons for failed renewal (too may tokens, the token has actually been revoked manually, password change).

I have a separate project in Google cloud for IMAP backup with a single token. I also didn't touch the password.
Now I generated a new token, lets see if refresh will work.

**Comment by joeyates** [comment]
Hi @vitaliy-sk 

Any news on this? Is refresh working?

**Comment by vitaliy-sk** [comment]
It worked ~6 days then stopped. 
Probably the root cause is not token management, my previous backup failed due to folder permission issues (my fault).
I can assume that it refreshed the token but not saved the new one.

**Comment by yesrod** [comment]
Google, in their infinite wisdom, has limited authorization for "testing" projects to seven days. After seven days, the token expires and must be reauthorized.

https://support.google.com/cloud/answer/10311615?hl=en#zippy=%2Ctesting

**Comment by ReubenBTalbott** [comment]
Is there any fix for this? I'm trying to setup an auto email backup system for myself and a neighbor and we both use Gmail.

**Comment by joeyates** [comment]
I have reverted imap-backup's default behaviour with GMail accounts to use standard user+password authentication.
OAuth2 is not usable with automated access.

I've released version 3.4.0 with this change.

Note: you need to create an 'App password' to use this functionality, which in turn means you need to set up 2-factor authentication on your GMail account.
